### DISCUSSION TITLE
Lesson 12 - Can't write test for DSCMint function

### DISCUSSION THREAD

**Discussion by smahdi-dev** [Cyfrin/foundry-full-course-cu#1747] [Q&A] [answered]
I want to write a test for a situation in which user wants to mint DSCs more than collateral. 
the following code is mintDSC function and _revertIfHealthFactorIsBroken functions:
```solidity
    function mintDSC(uint256 amountDscToMint) public moreThanZero(amountDscToMint) {
        s_DSCMinted[msg.sender] += amountDscToMint;
        _revertIfHealthFactorIsBroken(msg.sender);

        bool minted = i_dsc.mint(msg.sender, amountDscToMint);
        if (!minted) {
            revert DSCEngine__MintFailed();
        }
    }

    function _revertIfHealthFactorIsBroken(address user) internal view {
        uint256 userHealthFactor = _healthFactor(user);
        if (userHealthFactor < MIN_HEALTH_FACTOR) {
            revert DSCEngine__BreaksHealthFactor(userHealthFactor);
        }
    }
```

the problem lies in `DSCEngine__BreaksHealthFactor(userHealthFactor);` custom error which has a parameter.
look at the following code showing my test:
```solidity
    function testMintRevertsIfHealthFactorIsBroken() public depositedCollateral {
        vm.startPrank(USER);
        vm.expectRevert(abi.encodeWithSelector(DSCEngine.DSCEngine__BreaksHealthFactor, ??));
        engine.mintDSC(100000);
        vm.stopPrank();
    }
```

what value should i give to the custom error in `vm.expectRevert(abi.encodeWithSelector(DSCEngine.DSCEngine__BreaksHealthFactor, ??));` instead of `??`. the problem is that I have to call `getHealthFactor` after `engine.mintDSC(100000);` to calculate changes but I need what `getHealthFactor` returns before calling expectRevert to fill `??` section.

or mkae me aware if i am doing it the wrong way.

**Comment by EngrPips** [comment]
Hello @smahdi-dev, You can just use `vm.expectRevert()` without trying to pass in the exact data you are expecting to revert. Or you can maybe have some utility function that would calculate a user's `health factor` based on how much `DSC` they want to mint without actually minting any `DSC`.

**Reply by smahdi-dev** [reply]
Hi @EngrPips , I did it as you said without passing data:
```solidity
    function testMintRevertsIfHealthFactorIsBroken() public depositedCollateral {
        vm.startPrank(USER);
        vm.expectRevert(DSCEngine.DSCEngine__BreaksHealthFactor.selector);
        engine.mintDSC(100000);
        vm.stopPrank();
    }
```

but I get the following:
```
[FAIL. Reason: call did not revert as expected]
```

right now I'm stuck, can't get what's the problem

**Reply by EngrPips** [reply]
I meant, like, below 

```js
function testMintRevertsIfHealthFactorIsBroken() public depositedCollateral {
        vm.startPrank(USER);
        vm.expectRevert();
        engine.mintDSC(100000);
        vm.stopPrank();
    }
```

**Reply by smahdi-dev** [reply]
unfortunately, I'm getting the same error after doing this too 😢  @EngrPips 

I've read the code over and over but I can't find the issue bro

**Reply by EngrPips** [reply]
Hello @smahdi-dev, Below is my test set up which pass, The test is failing because the amount of `DSC` the user is minting does not break their health factor.

```js
uint256 wETH_test_amount = 1 ether;
uint256 inappropriate_DSC_to_mint = 0.6 ether;
```

```js
        modifier user_deposited_weth() {
        vm.startPrank(EngrPips);
        ERC20Mock(wETH_token_address).approve(address(dsc_engine), wETH_test_amount);
        dsc_engine.depositCollateral(wETH_token_address, wETH_test_amount);
        vm.stopPrank();
        _;
    }
```


```js
        function test_DSCEngine_revert_when_user_break_health_factor_when_they_mint() public user_deposited_weth {
        vm.prank(EngrPips);
        vm.expectRevert(DSCEngine.DSCEngine__health_factor_is_broken.selector);
        dsc_engine.mintDSC(inappropriate_DSC_to_mint);
    }
```

**Reply by EngrPips** [reply]
How much does the user deposits in your `depositedCollateral` modifier ?

**Reply by smahdi-dev** [reply]
Hi bro @EngrPips , I think I am doing the same. and i am depositing 10 ether. look:
```solidity
uint256 public constant AMOUNT_COLLATERAL = 10 ether;
```
```solidity
    modifier depositedCollateral() {
        vm.startPrank(USER);
        ERC20Mock(weth).approve(address(engine), AMOUNT_COLLATERAL);
        engine.depositCollateral(weth, AMOUNT_COLLATERAL);
        vm.stopPrank();
        _;
    }
```
```solidity
    function testMintRevertsIfHealthFactorIsBroken() public depositedCollateral {
        vm.startPrank(USER);
        vm.expectRevert();
        engine.mintDSC(10 ether);
        vm.stopPrank();
    }
```

**Reply by EngrPips** [reply]
If you run this test, you just showed me it should pass.

**Reply by smahdi-dev** [reply]
unfortunately, I'm still getting the error `[FAIL. Reason: call did not revert as expected]`.

could you please take a look at the code. https://github.com/smahdi-dev/defi-stablecoin

**Reply by EngrPips** [reply]
Alright, I will take a look and let you know whatever I find as soon as possible.

**Comment by EngrPips** [comment, marked as answer]
Hello @smahdi-dev, Below is your `_healthFactor` function. Why did you multiply `collateralAdjustedForThreshold` by `PRECISION` in the line below 

```js
return (collateralAdjustedForThreshold * PRECISION) / totalDscMinted;
```

```js
function _healthFactor(address user) internal view returns (uint256) {
        // total DSC minted
        // total collateral value
        (uint256 totalDscMinted, uint256 collateralValueInUsd) = _getAccountInformation(user);
        if (collateralValueInUsd == 0) {
            revert DSCEngine__NoCollateralDeposited();
        }
        if (totalDscMinted == 0) {
            revert DSCEngine__NoDscMinted();
        }
        uint256 collateralAdjustedForThreshold = (collateralValueInUsd * LIQUIDATION_THRESHOLD) / LIQUIDATION_PRECISION;
        return (collateralAdjustedForThreshold * PRECISION) / totalDscMinted;
    }
```
Below is mine 

```js
function _health_factor(address _user) internal view returns (uint256 health_factor) {
        (uint256 total_DSC_minted, uint256 collateral_value_in_USD) = _get_account_information(_user);
        console2.log("the total collateral value in usd is: ", collateral_value_in_USD);
        console2.log("the total DSC minted is: ", total_DSC_minted);
        if (total_DSC_minted == 0) return type(uint256).max;
        uint256 collateral_amount_for_threshold =
            (collateral_value_in_USD * LIQUIDATION_THRESHOLD) / LIQUIDATION_THRESHOLD_DIVIDER;
        console2.log("the collateral amount for threshold is: ", collateral_amount_for_threshold);
        health_factor = collateral_amount_for_threshold  / total_DSC_minted;
        console2.log("the calculated health factor is: ", health_factor);
        return health_factor;
    }
```

Maybe Patrick multiplied it in his codebase, and it worked for him, but when I ran into an issue and what the tutor did in the tutorial didn't work, I came back to thinking and figuring out how it would work for me. What a programmer does is figure out stuff.

**Reply by smahdi-dev** [reply]
@EngrPips , I removed `PRECISION` and it worked. 
as you said Patrick did it in the course.
and I also wanted to say that I have trouble understanding these math stuff because solidity doesn't support decimals, it always confuses me and I don't know when to add 1e18? sometimes Patrick adds 1e10 or 1e8.

how can i understand it better?

**Reply by EngrPips** [reply]
Hello @smahdi-dev, It is very typical that you find it difficult to understand solidity mathematics when you are just starting out with solidity, but with time, it will make more sense, and naturally, you will feel comfortable with it. I expect that you will run into another issue when you try to test that the `DSCEngine` properly mints `DSC` without the user breaking their `health_factor`. When that happens, make sure to reach out here.

**Reply by smahdi-dev** [reply]
thanks for the help ❤️ @EngrPips

**Reply by EngrPips** [reply]
I am glad I could help.
